control/controller.go

```diff
@@ -84,228 +69,202 @@ func New(c Config) *Controller {
+		c.finish(q)
+		c.countQueryRequest(q, labelCompileError)
+		return nil, q.Err()
 	}
-	return queries
-}
-
-// Shutdown will signal to the Controller that it should not accept any
-// new queries and that it should finish executing any existing queries.
-// This will return once the Controller's run loop has been exited and all
-// queries have been finished or until the Context has been canceled.
-func (c *Controller) Shutdown(ctx context.Context) error {
-	// Initiate the shutdown procedure by signaling to the run thread.
-	c.shutdown()
-
-	// Wait for the run loop to exit.
-	select {
-	case <-c.done:
-		return nil
-	case <-ctx.Done():
-		c.CancelAll()
-		return ctx.Err()
+	if err := c.enqueueQuery(q); err != nil {
+		q.setErr(err)
+		c.finish(q)
+		c.countQueryRequest(q, labelQueueError)
+		return nil, q.Err()
 	}
+	c.countQueryRequest(q, labelSuccess)
+	return q, nil
 }
 
-// CancelAll cancels all executing queries.
-func (c *Controller) CancelAll() {
+func (c *Controller) createQuery(ctx context.Context, ct flux.CompilerType) (*Query, error) {
 	c.queriesMu.RLock()
-	for _, q := range c.queries {
-		q.Cancel()
+	if c.shutdown {
+		c.queriesMu.RUnlock()
+		return nil, errors.New("query controller shutdown")
 	}
 	c.queriesMu.RUnlock()
-}
 
-func (c *Controller) run() {
-	defer close(c.done)
-
-	pq := newPriorityQueue()
-	for {
-		select {
-		// Wait for resources to free
-		case q := <-c.queryDone:
-			c.free(q)
-			c.queriesMu.Lock()
-			delete(c.queries, q.id)
-			c.queriesMu.Unlock()
-		// Wait for new queries
-		case q := <-c.newQueries:
-			pq.Push(q)
-			c.queriesMu.Lock()
-			c.queries[q.id] = q
-			c.queriesMu.Unlock()
-		// Wait for cancel query requests
-		case id := <-c.cancelRequest:
-			c.queriesMu.RLock()
-			q := c.queries[id]
-			c.queriesMu.RUnlock()
-			q.Cancel()
-		// Check if we have been signaled to shutdown.
-		case <-c.shutdownCtx.Done():
-			// We have been signaled to shutdown so drain the queues
-			// and exit the for loop.
-			c.drain(pq)
-			return
+	id := c.nextID()
+	labelValues := make([]string, len(c.labelKeys))
+	compileLabelValues := make([]string, len(c.labelKeys)+1)
+	for i, k := range c.labelKeys {
+		value := ctx.Value(k)
+		var str string
+		switch v := value.(type) {
+		case string:
+			str = v
+		case fmt.Stringer:
+			str = v.String()
 		}
+		labelValues[i] = str
+		compileLabelValues[i] = str
+	}
+	compileLabelValues[len(compileLabelValues)-1] = string(ct)
+
+	cctx, cancel := context.WithCancel(ctx)
+	parentSpan, parentCtx := StartSpanFromContext(
+		cctx,
+		"all",
+		c.metrics.allDur.WithLabelValues(labelValues...),
+		c.metrics.all.WithLabelValues(labelValues...),
+	)
+	q := &Query{
+		id:                 id,
+		labelValues:        labelValues,
+		compileLabelValues: compileLabelValues,
+		state:              Created,
+		c:                  c,
+		results:            make(chan flux.Result),
+		parentCtx:          parentCtx,
+		parentSpan:         parentSpan,
+		cancel:             cancel,
+	}
```

Comment:
Wouldn't it be necessary to pass `cctx` as `currentCtx` here?

Reply:
The `currentCtx` gets set by the state transitions. The `cctx` is passed as it is the parent of `parentCtx`.

Reply:
Oh great, I haven't noticed that! Thank you